pkg/authentication/authentication.go

```diff
@@ -134,21 +139,41 @@ func (a *Handler) handleTokenExchange(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	accountProvider, err := models.FindAccountProviderByProviderID("github", fmt.Sprintf("%d", githubUser.ID))
+	accountProvider, err := models.FindAccountProviderByProviderID(ProviderGitHub, fmt.Sprintf("%d", githubUser.ID))
 	if err != nil {
 		// If not found by provider ID, try to find by email
 		user, err := models.FindUserByEmail(githubUser.Email)
 		if err != nil {
-			log.Errorf("No existing account found for GitHub user %s (%s)", githubUser.Login, githubUser.Email)
-			http.Error(w, "No existing account found. Please sign up through the web interface first.", http.StatusNotFound)
-			return
-		}
+			// Check if there's an inactive user assigned by email
+			user, err = models.FindInactiveUserByEmail(githubUser.Email)
+			if err != nil {
+				log.Errorf("No existing account found for GitHub user %s (%s)", githubUser.Login, githubUser.Email)
+				http.Error(w, "No existing account found. Please sign up through the web interface first.", http.StatusNotFound)
+				return
+			}
 
-		accountProvider, err = user.GetAccountProvider("github")
-		if err != nil {
-			log.Errorf("User %s exists but has no GitHub account provider", githubUser.Email)
-			http.Error(w, "Account exists but GitHub provider not connected. Please connect GitHub through the web interface.", http.StatusNotFound)
-			return
+			accountProvider = &models.AccountProvider{
+				UserID:     user.ID,
+				Provider:   ProviderGitHub,
+				ProviderID: fmt.Sprintf("%d", githubUser.ID),
+				Username:   githubUser.Login,
+				Email:      githubUser.Email,
+				Name:       githubUser.Name,
+				AvatarURL:  githubUser.AvatarURL,
+			}
+
+			if err := accountProvider.Create(); err != nil {
+				log.Errorf("Failed to create account provider for inactive user: %v", err)
+				http.Error(w, "Internal server error", http.StatusInternalServerError)
+				return
+			}
+		} else {
+			accountProvider, err = user.GetAccountProvider(ProviderGitHub)
+			if err != nil {
+				log.Errorf("User %s exists but has no GitHub account provider", githubUser.Email)
+				http.Error(w, "Account exists but GitHub provider not connected. Please connect GitHub through the web interface.", http.StatusNotFound)
+				return
+			}
 		}
 	}
```

Comment:
The branching on this part makes it hard to follow. Can we move this logic to a `findOrCreateAccountProvider` function?